lib/maths/analytics/CBoostedTreeFactory.cc

```diff
@@ -187,6 +196,19 @@ CBoostedTreeFactory::buildForTrain(core::CDataFrame& frame, std::size_t dependen
     if (m_TreeImpl->m_Encoder->numberEncodedColumns() > 0) {
         this->initializeHyperparameters(frame);
         m_TreeImpl->m_Hyperparameters.initializeSearch();
+        if (m_TreeImpl->m_Hyperparameters.earlyStoppingEnabled()) {
+            // Add information about observed training runs to the GP
+            for (auto& hyperparameterLoss : *m_HyperparametersLosses) {
+                auto parameters = hyperparameterLoss.first.selectParametersVector(
+                    m_TreeImpl->m_Hyperparameters.tunableHyperparameters());
+                m_TreeImpl->m_Hyperparameters.addObservation(
+                    parameters, hyperparameterLoss.second, 0.0, true);
+            }
+            if (m_TreeImpl->m_Hyperparameters.stopEarly() == false) {
+                m_TreeImpl->m_Hyperparameters.resetSearch();
+            }
+            m_TreeImpl->m_Hyperparameters.resetBayesianOptimization();
+        }
```

Comment:
This just seems better placed in `CBoostedTreeHyperparameters::initializeSearch()`. This doesn't really seem part of the factory responsibilities. You then also don't need to add anything to `CBoostedTreeHyperparameters` public interface related to this change.

Reply:
I want this logic to be executed for initial training and not for incremental training (since parameter tuning for incremental training is negligible). By moving it to `CBoostedTreeHyperparameters::initializeSearch()` IMO it becomes a side-effect, hidden in the implementation. I can see your point since the logic operators on `m_Hyperparameters` exclusively, and so it would reduce coupling to refactor the logic into `CBoostedTreeHyperparameters`.

Reply:
Thanks!